File: python/39. Combination Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        candidates = sorted(candidates)
        self.helper(candidates, target, res, [])
        return res
    def helper(self, candidates, target, res, temp):
        if target == 0:
            res.append(temp)
            return
        for item in candidates:
            if item > target:
                break
            if temp and item < temp[-1]:
                continue
            else:
                self.helper(candidates, target - item, res, temp + [item])

File: cpp/1-50/39. Combination Sum.cpp
class Solution {
public:
	vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
		vector<vector<int>> res;
		if(candidates.empty())
			return res;
		vector<int> path;
		helper(candidates, 0, target, path, res, -1);
		return res;
	}
	void helper(vector<int> &candidates, int sum, int target, vector<int> &path, vector<vector<int>> &res, int index)
	{
		if (sum > target)
			return;
		if (sum == target)
		{
			res.push_back(path);
			return;
		}
		for(int i = 0; i < candidates.size(); i++)
		{
		    if(index >=0 && path[index] > candidates[i])
		        continue;
			path.push_back(candidates[i]);
			helper(candidates, sum + candidates[i], target, path, res, path.size() - 1);
			path.pop_back();
		}
	}
};